src/libultra_code/contpfs.c

```diff
@@ -0,0 +1,318 @@
+            ret = __osRepairPackId(pfs, id, &newid);
+            if (ret) {
+                return ret;
+            }
+            id = &newid;
+        } else if (ret != 0) {
+            return ret;
+        }
+    }
+
+    if ((id->deviceid & 0x01) == 0) {
+        ret = __osRepairPackId(pfs, id, &newid);
+        if (ret) {
+            return ret;
+        }
+        id = &newid;
+        if ((id->deviceid & 0x01) == 0) {
+            return PFS_ERR_DEVICE;
+        }
+    }
+
+    bcopy(id, pfs->id, BLOCKSIZE);
+
+    pfs->version = id->version;
+
+    pfs->banks = id->banks;
+    pfs->inodeStartPage = 1 + DEF_DIR_PAGES + (2 * pfs->banks);
+    pfs->dir_size = DEF_DIR_PAGES * PFS_ONE_PAGE;
+    pfs->inode_table = 1 * PFS_ONE_PAGE;
+    pfs->minode_table = (1 + pfs->banks) * PFS_ONE_PAGE;
+    pfs->dir_table = pfs->minode_table + (pfs->banks * PFS_ONE_PAGE);
+    if ((ret = __osContRamRead(pfs->queue, pfs->channel, PFS_LABEL_AREA, pfs->label)) != 0) {
+        return ret;
+    }
+
+    return 0;
+}
+#else
+#pragma GLOBAL_ASM("asm/non_matchings/code/contpfs/__osGetId.s")
+#endif
+
+s32 __osCheckId(OSPfs* pfs) {
+    u8 temp[BLOCKSIZE];
+    s32 ret;
+
+    if (pfs->activebank != 0) {
+        ret = __osPfsSelectBank(pfs, 0);
+        if (ret == PFS_ERR_NEW_PACK) {
+            ret = __osPfsSelectBank(pfs, 0);
+        }
+        if (ret != 0) {
+            return ret;
+        }
+    }
+
+    if ((ret = __osContRamRead(pfs->queue, pfs->channel, PFS_ID_0AREA, temp)) != 0) {
+        if (ret != PFS_ERR_NEW_PACK) {
+            return ret;
+        }
+        if ((ret = __osContRamRead(pfs->queue, pfs->channel, PFS_ID_0AREA, temp)) != 0) {
+            return ret;
+        }
+    }
+
+    if (bcmp(pfs->id, temp, BLOCKSIZE) != 0) {
+        return PFS_ERR_NEW_PACK;
+    }
+
+    return 0;
+}
+
+s32 __osPfsRWInode(OSPfs* pfs, __OSInode* inode, u8 flag, u8 bank) {
+    u8 sum;
+    s32 j;
+    s32 ret;
+    s32 offset;
+    u8* addr;
+
+    if (flag == PFS_READ && bank == __osPfsInodeCacheBank && (pfs->channel == __osPfsInodeCacheChannel)) {
+        bcopy(&__osPfsInodeCache, inode, sizeof(__OSInode));
+        return 0;
+    }
+
+    if ((pfs->activebank != 0) && (ret = __osPfsSelectBank(pfs, 0)) != 0) {
+        return ret;
+    }
+
+    offset = ((bank > 0) ? 1 : pfs->inodeStartPage);
+
+    if (flag == PFS_WRITE) {
+        inode->inodePage[0].inode_t.page =
+            __osSumcalc(inode->inodePage + offset, (PFS_INODE_SIZE_PER_PAGE - offset) * 2);
+    }
+
+    for (j = 0; j < PFS_ONE_PAGE; j++) {
+        addr = (u8*)(((u8*)inode) + (j * BLOCKSIZE));
+        if (flag == PFS_WRITE) {
+            ret = __osContRamWrite(pfs->queue, pfs->channel, pfs->inode_table + (bank * PFS_ONE_PAGE) + j, addr,
+                                   0);
+            ret = __osContRamWrite(pfs->queue, pfs->channel, pfs->minode_table + (bank * PFS_ONE_PAGE) + j, addr,
+                                   0); // why is this called twice??
```

Comment:
this doesn't look like the exact same call to me so the comment in unnecessary imo.